sdr_soapy.c

```diff
@@ -0,0 +1,374 @@
+    }
+
+    SoapySDRKwargs result = SoapySDRDevice_getHardwareInfo(SOAPY.dev);
+    if (result.size > 0) {
+        for (size_t i = 0; i < result.size; ++i) {
+            printf("%s=%s", result.keys[i], result.vals[i]);
+            if (i+1 < result.size) printf(", ");
+        }
+        printf("\n");
+    }
+    SoapySDRKwargs_clear(&result);
+
+    char* hw_key = SoapySDRDevice_getHardwareKey(SOAPY.dev);
+    if (hw_key) {
+        printf("soapy: hardware key is %s\n", hw_key);
+        SOAPY.hw_key = strdup(hw_key);
+        SoapySDR_free(hw_key);
+    }
+
+    if (SOAPY.channel) {
+        size_t supported_channels = SoapySDRDevice_getNumChannels(SOAPY.dev, SOAPY_SDR_RX);
+        if (SOAPY.channel >= supported_channels) {
+            fprintf(stderr, "soapy: device only supports %ld channels\n", supported_channels);
+            goto error;
+        }
+    }
+
+    length = 0;
+    char** available_antennas = SoapySDRDevice_listAntennas(SOAPY.dev, SOAPY_SDR_RX, SOAPY.channel, &length);
+    fprintf(stderr, "soapy: available antennas: ");
+    for (size_t i = 0; i < length; i++) {
+        fprintf(stderr, "%s", available_antennas[i]);
+        if (i+1 < length) printf(", ");
+    }
+    fprintf(stderr, "\n");
+    SoapySDRStrings_clear(&available_antennas, length);
+
+    if (SoapySDRDevice_setSampleRate(SOAPY.dev, SOAPY_SDR_RX, SOAPY.channel, Modes.sample_rate) != 0) {
+        fprintf(stderr, "soapy: setSampleRate failed: %s\n", SoapySDRDevice_lastError());
+        goto error;
+    }
+
+    if (SoapySDRDevice_setFrequency(SOAPY.dev, SOAPY_SDR_RX, SOAPY.channel, Modes.freq, NULL) != 0) {
+        fprintf(stderr, "soapy: setFrequency failed: %s\n", SoapySDRDevice_lastError());
+        goto error;
+    }
+
+    bool has_agc = SoapySDRDevice_hasGainMode(SOAPY.dev, SOAPY_SDR_RX, SOAPY.channel);
+    if (SOAPY.enable_agc) {
+        if (has_agc) {
+            if (SoapySDRDevice_setGainMode(SOAPY.dev, SOAPY_SDR_RX, SOAPY.channel, 1) != 0) {
+                fprintf(stderr, "soapy: setGainMode failed: %s\n", SoapySDRDevice_lastError());
+                goto error;
+            }
+        }
+        else {
+            fprintf(stderr, "soapy: device does not support enabling AGC\n");
+            goto error;
+        }
+    }
+    else if(has_agc) {
+        // by default, disable AGC
+        if (SoapySDRDevice_setGainMode(SOAPY.dev, SOAPY_SDR_RX, SOAPY.channel, 0) != 0) {
+            fprintf(stderr, "soapy: setGainMode failed: %s\n", SoapySDRDevice_lastError());
+            goto error;
+        }
+    }
+
+    if (Modes.gain != MODES_DEFAULT_GAIN) {
+        double gain;
+        if (SOAPY.dev_sdrplay) {
+            gain = Modes.gain;
+        } else {
+            gain = Modes.gain / 10.0;
+        }
+        if (SoapySDRDevice_setGain(SOAPY.dev, SOAPY_SDR_RX, SOAPY.channel, gain) != 0) {
+            fprintf(stderr, "soapy: setGain failed: %s\n", SoapySDRDevice_lastError());
+            goto error;
+        }
+    }
```

Comment:
This should probably unconditionally use Modes.gain directly with no division.

In older dump1090 versions, Modes.gain was an integer in tenths-of-a-dB. In more recent versions, it's just directly a floating-point gain in dB. I assume that SoapySDR wants gain in floating-point-dB (haven't looked at that in a while, but that's what I recall anyway)

Reply:
Yes, it's a floating point.  I've removed the divide-by-10 but more on this in my reply regarding the gain API implementation.